File: Python3 Solutions/3. Longest Substring Without Repeating Characters.py
# Question Link: https://leetcode.com/problems/longest-substring-without-repeating-characters/
from collections import defaultdict

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        letter_counts_map = defaultdict(int)
        left, right, length, longest_length_found = 0, 0, len(s), 0
        
        for right in range(length):
            letter_counts_map[s[right]] += 1
            while letter_counts_map[s[right]] > 1:
                letter_counts_map[s[left]] -= 1
                left += 1
                
            longest_length_found = max(longest_length_found, right - left + 1)
            
        return longest_length_found

File: Camp Progress/3. Longest Substring Without Repeating Characters.cpp
// Question Link: https://leetcode.com/problems/longest-substring-without-repeating-characters/

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int left = 0, answer = 0;
        unordered_map<char, int> countMap;
        for (int right = 0; right < s.size(); right++) {
            countMap[s[right]]++;
            while (countMap[s[right]] > 1) {
                countMap[s[left]]--;
                left++;
            }
            answer = max(answer, right - left + 1);
        }
        
        return answer;
    }
};